src/Distributed/distributed_fft_based_poisson_solver.jl

```diff
@@ -27,28 +29,35 @@ function DistributedFFTBasedPoissonSolver(arch, full_grid, local_grid)
     plan = PencilFFTs.PencilFFTPlan(size(full_grid), transform, proc_dims, MPI.COMM_WORLD)
     storage = PencilFFTs.allocate_input(plan)
 
-    return DistributedFFTBasedPoissonSolver(plan, full_grid, local_grid, eigenvalues, storage)
+    return DistributedFFTBasedPoissonSolver(arch, plan, full_grid, local_grid, eigenvalues, storage)
 end
 
-function solve_poisson_equation!(solver::DistributedFFTBasedPoissonSolver)
+function solve!(x, solver::DistributedFFTBasedPoissonSolver)
+    arch = solver.architecture
     λx, λy, λz = solver.eigenvalues
 
     # Apply forward transforms.
     solver.plan * solver.storage
 
-    # Solve the discrete Poisson equation.
-    RHS = ϕ = solver.storage[2]
-    @. ϕ = - RHS / (λx + λy + λz)
+    # Solve the discrete Poisson equation, storing the solution
+    # temporarily in xc and later extracting the real part into 
+    # the solution, x.
+    xc = b = solver.storage[2]
+    @. xc = - b / (λx + λy + λz)
 
     # Setting DC component of the solution (the mean) to be zero. This is also
     # necessary because the source term to the Poisson equation has zero mean
     # and so the DC component comes out to be ∞.
     if MPI.Comm_rank(MPI.COMM_WORLD) == 0
-        ϕ[1, 1, 1] = 0
+        xc[1, 1, 1] = 0
     end
 
     # Apply backward transforms.
     solver.plan \ solver.storage
+    xc_transposed = first(solver.storage)
+	
+    copy_event = launch!(arch, solver.my_grid, :xyz, copy_real_component!, x, xc_transposed, dependencies=device_event(arch))
+    wait(device(arch), copy_event)
 
-    return nothing
+    return x
```

Comment:
It modifies `x` and then returns it. We aren't good about this, but it's standard functionality for things like `ldiv!` and `mean!`. It's nice because then the result of the computation is displayed in the REPL, eg

```julia
julia> using Statistics

julia> a = rand(3)
3-element Vector{Float64}:
 0.8140284404676181
 0.04817379416233547
 0.5790117073593584

julia> b = rand(1)
1-element Vector{Float64}:
 0.5934424191746699

julia> mean!(b, a)
1-element Vector{Float64}:
 0.4804046473297706
```

see 

https://github.com/JuliaLang/Statistics.jl/blob/54f9b0d999813aa9fab039f632df222ffd2a96a8/src/Statistics.jl#L130

or

```julia
function mean!(R::AbstractArray, A::AbstractArray)
    sum!(R, A; init=true)
    x = max(1, length(R)) // length(A)
    R .= R .* x
    return R
end
```